Rename `acting_domain` to `acting_group` and some improvements for group cosets

ThomasBreuer opened the pull request:
- access defining data via functions not fields
- rename `acting_domain` to `acting_group`
- improve documentation
- `in` for left/right cosets now delegates to a membership test in a group
- iterator for left/right cosets now uses a group iterator

addresses #4283 and #4289

ThomasBreuer committed:
improvements for cosets

- access defining data via functions not fields
- rename `acting_domain` to `acting_group`
- improve documentation
- `in` for left/right cosets now delegates to a membership test in a group
- iterator for left/right cosets now uses a group iterator

fingolfin commented:
@ThomasBreuer does it resolve those PRs completely? Then you can write `resolves #X and resolves #Y` to make sure merging this PR automatically closes those issues. Alternatively you can manually link them via the web view
<img width="375" alt="Screen Shot 2024-11-13 at 08 10 38" src="https://github.com/user-attachments/assets/b82c17bb-24f4-4413-88a4-4fb3e6c54098">

ThomasBreuer commented:
> does it resolve those PRs completely

No.

lgoettgens commented on src/deprecations.jl:
```diff
@@ -147,3 +147,6 @@ Base.@deprecate_binding in_linear_system is_in_linear_system
 @deprecate has_minimal_generating_set(G::GAPGroup) has_minimal_size_generating_set(G)
 @deprecate set_minimal_generating_set(G::GAPGroup, v) set_minimal_size_generating_set(G, v)
+
+@deprecate acting_domain(C::GroupCoset) acting_group(C)
```
```suggestion
# deprecated for 1.3
@deprecate acting_domain(C::GroupCoset) acting_group(C)
```
just so we keep this information for the future

ThomasBreuer committed:
address a comment

ThomasBreuer committed:
change parameterization of the `GroupCoset` type

Add the type of the *subgroup* as a parameter,
then we can prescribe a better `Base.IteratorSize(::Type{<:GroupCoset})`,
as proposed in #4289.

Note that in principle, we could omit the type of the *big group* from the
parameters since it is the `parent_type` of the element type parameter.
(The design of the `GroupCoset` type dates back to the times when
we thought that a group has the same type as its subgroups.
At the time when this idea was given up, I should have changed
`GroupCoset` to take the *subgroup* type as a parameter.)

ThomasBreuer committed:
make `@inferred` happy

This was tricky:
The error messages were misleading, the problem occurred on a lower level.

ThomasBreuer commented:
Run tests / test (1.6, short, ubuntu-latest) failed with a segmentation fault, restarted.

fingolfin commented on src/Groups/cosets.jl:
```diff
@@ -1,20 +1,35 @@
-   function GroupCoset(G::T, H::GAPGroup, representative::S, side::Symbol) where {T<: GAPGroup, S<:GAPGroupElem}
-     return new{T, S}(G, H, representative, side, Ref{GapObj}())
+   function GroupCoset(G::TG, H::TH, representative::S, side::Symbol) where {TG <: GAPGroup, TH <: GAPGroup, S <:GAPGroupElem}
+     return new{TG, TH, S}(G, H, representative, side, Ref{GapObj}())
```
Maybe enforce the type relation(s) here, e.g. something like `@assert S == elem_type(TG)` or so (assuming that's correct)
